contracts/StablecoinConverter.sol

```diff
@@ -116,4 +119,147 @@ contract StablecoinConverter is EpochTokenLocker {
     function tokenIdToAddressMap(uint16 id) public view returns (address) {
         return IdToAddressBiMap.getAddressAt(registeredTokens, id);
     }
+
+    uint public previousSolutionBatchId;
+    TradeData[] public previousSolutionTrades;
+    uint[] public previousSolutionPrices;
+    uint16[] public previousSolutionTokenIdForPrice;
+
+    struct TradeData {
+        address owner;
+        uint volume;
+        uint16 orderId;
+    }
+
+    function submitSolution(
+        uint32 batchIndex,
+        address[] memory owner,  //tradeData is submitted as arrays
+        uint16[] memory orderId,
+        uint[] memory volume,
+        uint[] memory prices,
+        uint16[] memory tokenIdForPrice
+    ) public {
+        require(
+            batchIndex == getCurrentStateIndex() - 1,
+            "Solutions are no longer accepted for this batch"
+        );
+        // checkTradersUtility();
+        // checkPriceConformity(prices);
+        // checkNorm(prices);
+        int[] memory tokenConservation = new int[](prices.length);
+        undoPreviousSolution(batchIndex);
+        uint len = owner.length;
+        for(uint i = 0; i < len; i++){
+            Order memory order = orders[owner[i]][orderId[i]];
+            require(
+                tokenIdForPrice[findPriceIndex(order.sellToken, tokenIdForPrice)] == order.sellToken,
+                "sellTokenPriceIndex of order is incorrect"
+            );
+            require(
+                tokenIdForPrice[findPriceIndex(order.buyToken, tokenIdForPrice)] == order.buyToken,
+                "BuyTokenPriceId of order is incorrect"
+            );
+            require(order.validFrom <= batchIndex, "Order is not yet valid");
+            require(order.validUntil >= batchIndex, "Order is no longer valid");
+            require(
+                prices[order.sellToken].mul(order.buyAmount) >= prices[order.buyToken].mul(order.sellAmount),
+                "limit price not met"
+            );
+            // Assume for now that we always have sellOrders if (!order.isSellOrder) {
+            uint sellVolume = volume[i];
+            uint buyVolume = volume[i].mul(prices[findPriceIndex(order.buyToken, tokenIdForPrice)]) /
+                prices[findPriceIndex(order.sellToken, tokenIdForPrice)];
+            require(order.sellAmount >= sellVolume, "sellVolume bigger than specified in order");
+            orders[owner[i]][orderId[i]].buyAmount = uint128(
```

Comment:
Not sure where this mix of integer width comes from but should the volumes than also be provided in u128?